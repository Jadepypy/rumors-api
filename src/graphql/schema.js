import { GraphQLObjectType, GraphQLSchema } from 'graphql';

// Get individual objects
import GetArticle from './queries/GetArticle';
import GetReply from './queries/GetReply';
import GetUser from './queries/GetUser';
import GetCategory from './queries/GetCategory';
import ListArticles from './queries/ListArticles';
import ListReplies from './queries/ListReplies';
import ListCategories from './queries/ListCategories';
import ListArticleReplyFeedbacks from './queries/ListArticleReplyFeedbacks';

// Set individual objects
import CreateArticle from './mutations/CreateArticle';
import CreateReply from './mutations/CreateReply';
import CreateArticleReply from './mutations/CreateArticleReply';
import CreateCategory from './mutations/CreateCategory';
import CreateArticleCategory from './mutations/CreateArticleCategory';
import CreateOrUpdateArticleReplyFeedback from './mutations/CreateOrUpdateArticleReplyFeedback';
import CreateOrUpdateReplyRequestFeedback from './mutations/CreateOrUpdateReplyRequestFeedback';
import CreateOrUpdateArticleCategoryFeedback from './mutations/CreateOrUpdateArticleCategoryFeedback';
import CreateOrUpdateReplyRequest from './mutations/CreateOrUpdateReplyRequest';
import CreateOrUpdateUser from './mutations/CreateOrUpdateUser';
import UpdateArticleReplyStatus from './mutations/UpdateArticleReplyStatus';
import UpdateArticleCategoryStatus from './mutations/UpdateArticleCategoryStatus';
import UpdateUser from './mutations/UpdateUser';

export default new GraphQLSchema({
  query: new GraphQLObjectType({
    name: 'Query',
    fields: {
      GetArticle,
      GetReply,
      GetUser,
      GetCategory,
      ListArticles,
      ListReplies,
      ListCategories,
      ListArticleReplyFeedbacks,
    },
  }),
  mutation: new GraphQLObjectType({
    name: 'Mutation',
    fields: {
      CreateArticle,
      CreateReply,
      CreateArticleReply,
      CreateCategory,
      CreateArticleCategory,
      CreateReplyRequest: {
        ...CreateOrUpdateReplyRequest,
        deprecationReason: 'Use CreateOrUpdateReplyRequest instead',
      },
      CreateOrUpdateReplyRequest,
      CreateOrUpdateArticleReplyFeedback,
      CreateOrUpdateArticleCategoryFeedback,
      CreateOrUpdateReplyRequestFeedback,
      CreateOrUpdateUser,
      UpdateArticleReplyStatus,
      UpdateArticleCategoryStatus,
      UpdateUser,
    },
  }),
});
