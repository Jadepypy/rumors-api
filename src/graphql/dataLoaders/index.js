import docLoaderFactory from './docLoaderFactory';
import analyticsLoaderFactory from './analyticsLoaderFactory';
import articleRepliesByReplyIdLoaderFactory from './articleRepliesByReplyIdLoaderFactory';
import articleCategoriesByCategoryIdLoaderFactory from './articleCategoriesByCategoryIdLoaderFactory';
import articleReplyFeedbacksLoaderFactory from './articleReplyFeedbacksLoaderFactory';
import articleCategoryFeedbacksLoaderFactory from './articleCategoryFeedbacksLoaderFactory';
import searchResultLoaderFactory from './searchResultLoaderFactory';
import urlLoaderFactory from './urlLoaderFactory';
import repliedArticleCountLoaderFactory from './repliedArticleCountLoaderFactory';
import votedArticleReplyCountLoaderFactory from './votedArticleReplyCountLoaderFactory';
import userLevelLoaderFactory from './userLevelLoaderFactory';
import userLoaderFactory from './userLoaderFactory';

export default class DataLoaders {
  // List of data loaders
  //
  get docLoader() {
    return this._checkOrSetLoader('docLoader', docLoaderFactory);
  }
  get articleRepliesByReplyIdLoader() {
    return this._checkOrSetLoader(
      'articleRepliesByReplyIdLoader',
      articleRepliesByReplyIdLoaderFactory
    );
  }
  get articleCategoriesByCategoryIdLoader() {
    return this._checkOrSetLoader(
      'articleCategoriesByCategoryIdLoader',
      articleCategoriesByCategoryIdLoaderFactory
    );
  }
  get articleReplyFeedbacksLoader() {
    return this._checkOrSetLoader(
      'articleReplyFeedbacksLoader',
      articleReplyFeedbacksLoaderFactory
    );
  }
  get articleCategoryFeedbacksLoader() {
    return this._checkOrSetLoader(
      'articleCategoryFeedbacksLoader',
      articleCategoryFeedbacksLoaderFactory
    );
  }
  get searchResultLoader() {
    return this._checkOrSetLoader(
      'searchResultLoader',
      searchResultLoaderFactory
    );
  }

  get urlLoader() {
    return this._checkOrSetLoader('urlLoader', urlLoaderFactory);
  }

  get repliedArticleCountLoader() {
    return this._checkOrSetLoader(
      'repliedArticleCountLoader',
      repliedArticleCountLoaderFactory
    );
  }

  get votedArticleReplyCountLoader() {
    return this._checkOrSetLoader(
      'votedArticleReplyCountLoader',
      votedArticleReplyCountLoaderFactory
    );
  }

  get userLoader() {
    return this._checkOrSetLoader('userLoader', userLoaderFactory);
  }

  get userLevelLoader() {
    return this._checkOrSetLoader('userLevelLoader', userLevelLoaderFactory);
  }

  get analyticsLoader() {
    return this._checkOrSetLoader('analyticsLoader', analyticsLoaderFactory);
  }

  // inner-workings
  //
  constructor() {
    this._loaders = {};
  }

  _checkOrSetLoader(name, factoryFn) {
    if (this._loaders[name]) return this._loaders[name];

    this._loaders[name] = factoryFn(this);
    return this._loaders[name];
  }
}
