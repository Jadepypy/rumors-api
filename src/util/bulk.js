const BATCH_SIZE = 100;

export default class Bulk {
  constructor(client, batchSize = BATCH_SIZE) {
    this.client = client;
    this.batchSize = batchSize;
    this._operations = [];
    this.actionsCount = 0;
  }

  async push(items, count = 1) {
    this._operations.push(...items);
    this.actionsCount += count;
    if (
      this.actionCounts > this.batchSize ||
      this._operations.length > 3 * this.batchSize
    ) {
      await this.flush();
    }
    return this;
  }

  async flush() {
    if (this._operations.length === 0) {
      return this;
    }

    // Process entires queued up
    const op = this._operations.splice(0);
    this.actionCounts = 0;
    const { body } = await this.client.bulk({
      body: op,
      refresh: 'true',
    });
    if (body.errors) {
      console.error(body.errors);
    } else {
      console.info(`${body.items.length} items processed`);
    }

    return this;
  }
}
