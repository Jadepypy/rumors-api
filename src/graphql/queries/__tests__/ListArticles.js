import gql from 'util/GraphQL';
import { loadFixtures, unloadFixtures } from 'util/fixtures';
import { getCursor } from 'graphql/util';
import fixtures from '../__fixtures__/ListArticles';
import fetch from 'node-fetch';
import { imageHash } from 'image-hash';

jest.mock('node-fetch');
jest.mock('image-hash', () => ({
  __esModule: true,
  imageHash: jest.fn(),
}));

describe('ListArticles', () => {
  beforeAll(() => loadFixtures(fixtures));

  it('lists all articles', async () => {
    expect(
      await gql`
        {
          ListArticles {
            totalCount
            edges {
              node {
                id
              }
              cursor
            }
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('sorts', async () => {
    expect(
      await gql`
        {
          ListArticles(orderBy: [{ updatedAt: DESC }]) {
            edges {
              node {
                id
                updatedAt
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot('by updatedAt DESC');

    expect(
      await gql`
        {
          ListArticles(orderBy: [{ replyRequestCount: DESC }]) {
            edges {
              node {
                id
                replyRequestCount
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot('by replyRequestCount DESC');

    expect(
      await gql`
        {
          ListArticles(orderBy: [{ lastRepliedAt: DESC }]) {
            edges {
              node {
                id
                articleReplies(status: NORMAL) {
                  createdAt
                }
              }
            }
          }
        }
      `()
    ).toMatchSnapshot('by lastRepliedAt DESC');
  });

  const testReplyCount = async expression => {
    // Lists only articles with more than one reply.
    const pair = expression ? `${expression}: 1` : expression;
    expect(
      await gql`
        {
          ListArticles(filter: { replyCount: {${pair}} }) {
            edges {
              node {
                id
                replyCount
              }
            }
          }
        }
      `()
    ).toMatchSnapshot();
  };

  it('filters by replyCount EQ', () => testReplyCount('EQ'));
  it('filters by replyCount LT', () => testReplyCount('LT'));
  it('filters by replyCount GT', () => testReplyCount('GT'));
  it('filters by invalid operator', () => testReplyCount('INVALID'));
  it('filters by null operator', () => testReplyCount(''));

  it('filters by moreLikeThis', async () => {
    expect(
      await gql`
        {
          ListArticles(
            filter: {
              moreLikeThis: {
                like: "人間相見是何年？牽攣乖隔，各欲白首。"
                minimumShouldMatch: "5%"
              }
            }
          ) {
            edges {
              node {
                id
              }
              highlight {
                text
                hyperlinks {
                  url
                  title
                  summary
                }
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('filters by moreLikeThis and given text, find articles containing hyperlinks with the said text', async () => {
    expect(
      await gql`
        query($like: String) {
          ListArticles(
            filter: { moreLikeThis: { like: $like, minimumShouldMatch: "5%" } }
          ) {
            edges {
              node {
                id
                hyperlinks {
                  summary
                  topImageUrl
                }
              }
              highlight {
                text
                hyperlinks {
                  url
                  title
                  summary
                }
              }
            }
          }
        }
      `({
        like: `
          1. text -> ariticles linked to the content
          居有頃，倚柱彈其劍，歌曰：「長鋏歸來乎！食無魚！」左右以告。孟嘗君曰：「食之
          ，比門下之客。」居有頃，復彈其鋏，歌曰：「長鋏歸來乎！出無車！」
        `,
      })
    ).toMatchSnapshot();
  });

  it("filters by moreLikeThis and given URL, find articles with the said URL's content", async () => {
    expect(
      await gql`
        query($like: String) {
          ListArticles(
            filter: { moreLikeThis: { like: $like, minimumShouldMatch: "5%" } }
          ) {
            edges {
              node {
                id
                hyperlinks {
                  summary
                  topImageUrl
                }
              }
              highlight {
                text
                hyperlinks {
                  url
                  title
                  summary
                }
              }
            }
          }
        }
      `({
        like: `
          2. URL -> article with given URL's content
          http://出師表.com
        `,
      })
    ).toMatchSnapshot();
  });

  it('filters by replyRequestCount', async () => {
    // Lists only articles with more than 1 reply requests
    expect(
      await gql`
        {
          ListArticles(filter: { replyRequestCount: { GT: 1 } }) {
            edges {
              node {
                id
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('filters by userId, appId and fromUserOfArticleId', async () => {
    // Lists only articles by userId & appId
    expect(
      await gql`
        {
          ListArticles(filter: { userId: "user1", appId: "app1" }) {
            edges {
              node {
                id
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot('userId = user1, appId = app1');

    // Lists only articles by fromUserOfArticleId
    expect(
      await gql`
        {
          ListArticles(filter: { fromUserOfArticleId: "listArticleTest1" }) {
            edges {
              node {
                id
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot('author of listArticleTest1');
  });

  it('filters by time range', async () => {
    expect(
      await gql`
        {
          ListArticles(
            filter: { createdAt: { GT: "2020-02-06T00:00:00.000Z" } }
          ) {
            edges {
              node {
                id
                createdAt
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot('later than 2020-02-06');
    expect(
      await gql`
        {
          ListArticles(
            filter: { createdAt: { LTE: "2020-02-06T00:00:00.000Z" } }
          ) {
            edges {
              node {
                id
                createdAt
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot('earlier or equal to 2020-02-06');
    expect(
      await gql`
        {
          ListArticles(
            filter: {
              createdAt: {
                GTE: "2020-02-04T00:00:00.000Z"
                LTE: "2020-02-06T00:00:00.000Z"
              }
            }
          ) {
            edges {
              node {
                id
                createdAt
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot('between 2020-02-04 and 2020-02-06');
  });

  it('filters by replies time range', async () => {
    expect(
      await gql`
        {
          ListArticles(
            filter: { repliedAt: { GT: "2020-02-06T00:00:00.000Z" } }
          ) {
            edges {
              node {
                id
                articleReplies(status: NORMAL) {
                  createdAt
                }
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot('later than 2020-02-06');
    expect(
      await gql`
        {
          ListArticles(
            filter: { repliedAt: { LTE: "2020-02-06T00:00:00.000Z" } }
          ) {
            edges {
              node {
                id
                articleReplies(status: NORMAL) {
                  createdAt
                }
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot('earlier or equal to 2020-02-06');
    expect(
      await gql`
        {
          ListArticles(
            filter: {
              repliedAt: {
                GTE: "2020-02-04T00:00:00.000Z"
                LTE: "2020-02-06T00:00:00.000Z"
              }
            }
          ) {
            edges {
              node {
                id
                articleReplies(status: NORMAL) {
                  createdAt
                }
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot('between 2020-02-04 and 2020-02-06');
  });

  it('filters by mixed query', async () => {
    // Mixes 'should' and 'filter' query. At least 1 'should' must match.
    // Therefore, this query should only match 1 result instead of all that satisfies replyRequestCount: { GT: 0 }
    expect(
      await gql`
        {
          ListArticles(
            filter: {
              moreLikeThis: {
                like: "憶昔封書與君夜，金鑾殿後欲明天。今夜封書在何處？廬山庵裏曉燈前。籠鳥檻猿俱未死，人間相見是何年？"
              }
              replyRequestCount: { GT: 0 }
            }
          ) {
            edges {
              node {
                id
              }
              highlight {
                text
                hyperlinks {
                  url
                  title
                  summary
                }
              }
            }
            totalCount
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('throws error when author filter is not set correctly', async () => {
    const { errors: notExistError } = await gql`
      {
        ListArticles(filter: { fromUserOfArticleId: "not-exist" }) {
          edges {
            node {
              id
            }
          }
          totalCount
        }
      }
    `();
    expect(notExistError).toMatchSnapshot();
  });

  it('supports after', async () => {
    expect(
      await gql`
        query($cursor: String) {
          ListArticles(after: $cursor) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `({ cursor: getCursor(['listArticleTest2']) })
    ).toMatchSnapshot();
  });

  it('supports before', async () => {
    expect(
      await gql`
        query($cursor: String) {
          ListArticles(before: $cursor) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `({ cursor: getCursor(['listArticleTest2']) })
    ).toMatchSnapshot();
  });

  it('should fail if before and after both exist', async () => {
    expect(
      await gql`
        query($cursor: String) {
          ListArticles(before: $cursor, after: $cursor) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `({ cursor: getCursor(['listArticleTest2']) })
    ).toMatchSnapshot();
  });

  it('correctly handles empty lists without errors', async () => {
    expect(
      await gql`
        {
          ListArticles(
            filter: { moreLikeThis: { like: "ThisShouldNotExist" } }
          ) {
            edges {
              node {
                id
              }
              highlight {
                text
                hyperlinks {
                  url
                  title
                  summary
                }
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('use filter categoryIds to list articles', async () => {
    expect(
      await gql`
        {
          ListArticles(
            orderBy: [{ _score: DESC }]
            filter: { categoryIds: ["category1", "category-author-1"] }
          ) {
            edges {
              node {
                id
              }
            }
            totalCount
            pageInfo {
              firstCursor
              lastCursor
            }
          }
        }
      `()
    ).toMatchSnapshot();
  });

  it('filters via article reply feedback count', async () => {
    expect(
      await gql`
        {
          ListArticles(
            filter: { hasArticleReplyWithMorePositiveFeedback: true }
          ) {
            edges {
              node {
                id
                articleReplies(status: NORMAL) {
                  positiveFeedbackCount
                  negativeFeedbackCount
                }
              }
            }
          }
        }
      `()
    ).toMatchSnapshot('hasArticleReplyWithMorePositiveFeedback = true');

    expect(
      await gql`
        {
          ListArticles(
            filter: { hasArticleReplyWithMorePositiveFeedback: false }
          ) {
            edges {
              node {
                id
                articleReplies(status: NORMAL) {
                  positiveFeedbackCount
                  negativeFeedbackCount
                }
              }
            }
          }
        }
      `()
    ).toMatchSnapshot('hasArticleReplyWithMorePositiveFeedback = false');
  });

  it('filters via articleRepliesFrom', async () => {
    expect(
      await gql`
        {
          ListArticles(filter: { articleRepliesFrom: { userId: "user1" } }) {
            edges {
              node {
                id
                articleReplies(status: NORMAL) {
                  user {
                    id
                  }
                }
              }
            }
          }
        }
      `({}, { appId: 'WEBSITE' })
    ).toMatchSnapshot('has articleReply from user1');

    expect(
      await gql`
        {
          ListArticles(
            filter: { articleRepliesFrom: { userId: "user1", exists: false } }
          ) {
            edges {
              node {
                id
                articleReplies(status: NORMAL) {
                  user {
                    id
                  }
                }
              }
            }
          }
        }
      `({}, { appId: 'WEBSITE' })
    ).toMatchSnapshot('do not have articleReply from user1');
  });

  it('filters by reply types', async () => {
    expect(
      await gql`
        {
          ListArticles(filter: { replyTypes: [NOT_RUMOR, OPINIONATED] }) {
            edges {
              node {
                id
                articleReplies(status: NORMAL) {
                  replyType
                }
              }
            }
          }
        }
      `({}, { appId: 'WEBSITE' })
    ).toMatchSnapshot('replied with NOT_RUMOR and OPINIONATED');
  });

  it('returns activities stats', async () => {
    expect(
      await gql`
        {
          ListArticles {
            edges {
              node {
                id
                stats(dateRange: { GTE: "2020-01-03", LTE: "2020-01-05" }) {
                  date
                  webUser
                  webVisit
                  lineUser
                  lineVisit
                }
              }
            }
          }
        }
      `()
    ).toMatchSnapshot('articles with stats');
  });

  it('filters by article types', async () => {
    expect(
      await gql`
        {
          ListArticles(filter: { articleTypes: [IMAGE, AUDIO] }) {
            edges {
              node {
                id
                articleType
              }
            }
          }
        }
      `({}, { appId: 'WEBSITE' })
    ).toMatchSnapshot('IMAGE and AUDIO articles');
  });

  it('filters by mediaUrl, without MEDIA_ARTICLE_SUPPORT env', async () => {
    fetch.mockImplementation(() =>
      Promise.resolve({
        status: 200,
        body: {},
        buffer: jest.fn(),
      })
    );
    imageHash.mockImplementation((file, bits, method, callback) =>
      callback(undefined, 'ffff8000')
    );

    expect(
      await gql`
        {
          ListArticles(
            filter: { mediaUrl: "http://foo.com/input_image.jpeg" }
          ) {
            edges {
              node {
                id
                articleType
                attachmentUrl
                attachmentHash
              }
            }
          }
        }
      `({}, { appId: 'WEBSITE' })
    ).toMatchSnapshot('should return empty');
  });

  it('filters by mediaUrl with MEDIA_ARTICLE_SUPPORT env', async () => {
    process.env.MEDIA_ARTICLE_SUPPORT = true;
    fetch.mockImplementation(() =>
      Promise.resolve({
        status: 200,
        body: {},
        buffer: jest.fn(),
      })
    );
    imageHash.mockImplementation((file, bits, method, callback) =>
      callback(undefined, 'ffff8000')
    );

    expect(
      await gql`
        {
          ListArticles(
            filter: { mediaUrl: "http://foo.com/input_image.jpeg" }
          ) {
            edges {
              node {
                id
                articleType
                attachmentUrl
                attachmentHash
              }
            }
          }
        }
      `({}, { appId: 'WEBSITE' })
    ).toMatchSnapshot('attachmentHash should be ffff8000');
    delete process.env.MEDIA_ARTICLE_SUPPORT;
  });

  afterAll(() => unloadFixtures(fixtures));
});
