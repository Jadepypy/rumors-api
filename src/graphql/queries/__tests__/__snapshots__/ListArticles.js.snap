// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListArticles correctly handles empty lists without errors 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [],
      "pageInfo": Object {
        "firstCursor": null,
        "lastCursor": null,
      },
      "totalCount": 0,
    },
  },
}
`;

exports[`ListArticles filters by invalid operator 1`] = `
Object {
  "errors": Array [
    [GraphQLError: Field "INVALID" is not defined by type ListArticleReplyCountExpr.],
  ],
}
`;

exports[`ListArticles filters by mixed query 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "totalCount": 1,
    },
  },
}
`;

exports[`ListArticles filters by moreLikeThis 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJsaXN0QXJ0aWNsZVRlc3QzIl0=",
        "lastCursor": "WyJsaXN0QXJ0aWNsZVRlc3QxIl0=",
      },
      "totalCount": 2,
    },
  },
}
`;

exports[`ListArticles filters by moreLikeThis and given URL, find articles with the said URL's content 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "hyperlinks": null,
            "id": "listArticleTest2",
          },
        },
      ],
    },
  },
}
`;

exports[`ListArticles filters by moreLikeThis and given text, find articles containing hyperlinks with the said text 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "hyperlinks": Array [
              Object {
                "summary": "居有頃，倚柱彈其劍，歌曰：「長鋏歸來乎！食無魚。」左右以告。孟嘗君曰：「食之，比門下之客。」",
                "topImageUrl": "http://gohome.com/image.jpg",
              },
            ],
            "id": "listArticleTest4",
          },
        },
      ],
    },
  },
}
`;

exports[`ListArticles filters by null operator 1`] = `
Object {
  "data": Object {
    "ListArticles": null,
  },
  "errors": Array [
    [GraphQLError: Invalid Expression!],
  ],
}
`;

exports[`ListArticles filters by replies time range 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest4",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest2",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "totalCount": 3,
    },
  },
}
`;

exports[`ListArticles filters by replies time range 2`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "totalCount": 2,
    },
  },
}
`;

exports[`ListArticles filters by replies time range 3`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "totalCount": 2,
    },
  },
}
`;

exports[`ListArticles filters by replyCount EQ 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest2",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJsaXN0QXJ0aWNsZVRlc3QyIl0=",
        "lastCursor": "WyJsaXN0QXJ0aWNsZVRlc3QyIl0=",
      },
      "totalCount": 1,
    },
  },
}
`;

exports[`ListArticles filters by replyCount GT 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJsaXN0QXJ0aWNsZVRlc3QxIl0=",
        "lastCursor": "WyJsaXN0QXJ0aWNsZVRlc3QxIl0=",
      },
      "totalCount": 1,
    },
  },
}
`;

exports[`ListArticles filters by replyCount LT 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest4",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJsaXN0QXJ0aWNsZVRlc3Q0Il0=",
        "lastCursor": "WyJsaXN0QXJ0aWNsZVRlc3QzIl0=",
      },
      "totalCount": 2,
    },
  },
}
`;

exports[`ListArticles filters by replyRequestCount 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "totalCount": 1,
    },
  },
}
`;

exports[`ListArticles filters by time range 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest4",
          },
        },
      ],
      "totalCount": 1,
    },
  },
}
`;

exports[`ListArticles filters by time range 2`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest2",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "totalCount": 3,
    },
  },
}
`;

exports[`ListArticles filters by time range 3`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest2",
          },
        },
      ],
      "totalCount": 2,
    },
  },
}
`;

exports[`ListArticles filters by userId, appId and fromUserOfArticleId 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest2",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "totalCount": 2,
    },
  },
}
`;

exports[`ListArticles filters by userId, appId and fromUserOfArticleId 2`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest2",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "totalCount": 2,
    },
  },
}
`;

exports[`ListArticles lists all articles 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "cursor": "WyJsaXN0QXJ0aWNsZVRlc3Q0Il0=",
          "node": Object {
            "id": "listArticleTest4",
          },
        },
        Object {
          "cursor": "WyJsaXN0QXJ0aWNsZVRlc3QzIl0=",
          "node": Object {
            "id": "listArticleTest3",
          },
        },
        Object {
          "cursor": "WyJsaXN0QXJ0aWNsZVRlc3QyIl0=",
          "node": Object {
            "id": "listArticleTest2",
          },
        },
        Object {
          "cursor": "WyJsaXN0QXJ0aWNsZVRlc3QxIl0=",
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJsaXN0QXJ0aWNsZVRlc3Q0Il0=",
        "lastCursor": "WyJsaXN0QXJ0aWNsZVRlc3QxIl0=",
      },
      "totalCount": 4,
    },
  },
}
`;

exports[`ListArticles should fail if before and after both exist 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": null,
      "pageInfo": Object {
        "firstCursor": "WyJsaXN0QXJ0aWNsZVRlc3Q0Il0=",
        "lastCursor": "WyJsaXN0QXJ0aWNsZVRlc3QxIl0=",
      },
      "totalCount": 4,
    },
  },
  "errors": Array [
    [GraphQLError: Use of before & after is prohibited.],
  ],
}
`;

exports[`ListArticles sorts 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest4",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest2",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WzQsImxpc3RBcnRpY2xlVGVzdDQiXQ==",
        "lastCursor": "WzEsImxpc3RBcnRpY2xlVGVzdDEiXQ==",
      },
      "totalCount": 4,
    },
  },
}
`;

exports[`ListArticles sorts 2`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest2",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest4",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WzIsImxpc3RBcnRpY2xlVGVzdDEiXQ==",
        "lastCursor": "WzAsImxpc3RBcnRpY2xlVGVzdDMiXQ==",
      },
      "totalCount": 4,
    },
  },
}
`;

exports[`ListArticles sorts 3`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest4",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest2",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WzE1ODE0MzM4NjQ0NzIsImxpc3RBcnRpY2xlVGVzdDQiXQ==",
        "lastCursor": "WzE1ODA5MTU0NjQ0NzIsImxpc3RBcnRpY2xlVGVzdDMiXQ==",
      },
      "totalCount": 4,
    },
  },
}
`;

exports[`ListArticles supports after 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest1",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJsaXN0QXJ0aWNsZVRlc3Q0Il0=",
        "lastCursor": "WyJsaXN0QXJ0aWNsZVRlc3QxIl0=",
      },
      "totalCount": 4,
    },
  },
}
`;

exports[`ListArticles supports before 1`] = `
Object {
  "data": Object {
    "ListArticles": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "listArticleTest4",
          },
        },
        Object {
          "node": Object {
            "id": "listArticleTest3",
          },
        },
      ],
      "pageInfo": Object {
        "firstCursor": "WyJsaXN0QXJ0aWNsZVRlc3Q0Il0=",
        "lastCursor": "WyJsaXN0QXJ0aWNsZVRlc3QxIl0=",
      },
      "totalCount": 4,
    },
  },
}
`;

exports[`ListArticles throws error when author filter is not set correctly 1`] = `
Array [
  [GraphQLError: Both appId and userId must be specified at once],
]
`;

exports[`ListArticles throws error when author filter is not set correctly 2`] = `
Array [
  [GraphQLError: fromUserOfArticleId does not match any existing articles],
]
`;
